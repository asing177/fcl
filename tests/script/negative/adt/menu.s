type Size {
  Big;
  Small;
}

type Item {
  Chips(Size size);
  Milkshake(int millilitres, bool sprinkles);
  Salad;
}

global Bogus1 bog1 = Nope; // undefined constructor + undefined type
global Bogus2 bog2;        // undefined type

@initial
foo(int volume) {
  z = Milkshake(volume, false);
  y = Milkshake(volume, volume); // bad arg
  x = Milkshake(1, true, 1);     // bad arity
  w = Chips;                     // bad arity
  terminate()
}

@initial
bar(Item it, Bogus b) {           // undefined type
  x = case it {
    Milkshake(p) -> p;            // bad arity
    Milkshake(p1,p2) -> p2;       // branch body of wrong type
    Milkshake(p1,p1) -> p1;       // shadowing / duplicate variable patterns
    Milkshake(p1,p2,p3,p4) -> p1; // bad arity
    Milkshake(1, 2) -> 42;        // literal pattern of wrong type
    Chips(Big) -> 0;
    Salad() -> 0;
    Salad -> 0;                   // currently no reachability check
    Chips(Little) -> 1;           // undefined constructor
    _ -> 0
  };
  y = case x {
    0 -> "hello";
    _ -> "world"
  };
  terminate()
}
