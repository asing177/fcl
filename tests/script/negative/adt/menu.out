Error at Line 5, Column 15:
   Reference to undefined constructor:
      Nope

Error at Line 5, Column 15:
   Reference to undefined type:
      Bogus1

Error at Line 6, Column 8:
   Reference to undefined type:
      Bogus2

Type error at Line 11, Column 25:
   Expected: 'bool' inferred from type of field 'sprinkles' of constructor 'Milkshake' on Line 11, Column 25
   But got:  'int' inferred from the type signature of argument 1 of function 'foo' on Line 9, Column 9

Error at Line 12, Column 26:
   Constructor 'Milkshake' is applied to too many arguments ('1')

Error at Line 13, Column 7:
   Constructor 'Chips' is not applied to enough arguments; expecting an argument of type 'Size' for field 'size'

Error at Line 18, Column 20:
   Reference to undefined type:
      Bogus

Error at Line 20, Column 5:
   Constructor 'Milkshake' is not applied to enough arguments; expecting an argument of type 'bool' for field 'sprinkles'

Type error at Line 20, Column 21:
   Expected: 'int' inferred from variable binding 'p' in case body on Line 20, Column 15
   But got:  'bool' inferred from variable binding 'p2' in case body on Line 21, Column 18

Type error at Line 20, Column 21:
   Expected: 'int' inferred from variable binding 'p' in case body on Line 20, Column 15
   But got:  'bool' inferred from variable binding 'p1' in case body on Line 22, Column 18

Error at Line 22, Column 15:
   The binding 'p1' shadows the variable bound by a pattern
      of the same name, of type 'bool',
      bound at Line 22, Column 18.

Error at Line 23, Column 21:
   Constructor 'Milkshake' is applied to too many patterns ('p3')

Error at Line 23, Column 24:
   Constructor 'Milkshake' is applied to too many patterns ('p4')

Type error at Line 24, Column 18:
   Expected: 'bool' inferred from type of field 'sprinkles' of constructor 'Milkshake' on Line 24, Column 18
   But got:  'int' inferred from literal '2' on Line 24, Column 18

Error at Line 28, Column 11:
   Reference to undefined constructor:
      Little